### DISCUSSION TITLE
proposal for maximising code sharing between sqlalchemy and pydantic models

### DISCUSSION THREAD

**Discussion by sandys** [fastapi/fastapi#9046] [Questions] [answered]
We are in the initial days of using this in our production code (which is migrating from flask). Unsure if there will be dragons ahead, but wanted to put it here for the community

This may be related to other similar bugs - https://github.com/tiangolo/fastapi/issues/214 and 

we are basically leveraging the "classic mapping" method of sqlalchemy https://docs.sqlalchemy.org/en/13/orm/mapping_styles.html#classical-mappings

The important thing is to not use declarative_base. This allows us to basically develop a pydantic basemodel and derive sqlalchemy models from that. 
Ideally i would have wanted support from pydantic to define a pure python class and be able to create a basemodel out of that. But I dont think that's possible.

```

from sqlalchemy import Table, MetaData, Column, Integer, String, ForeignKey
from sqlalchemy.orm import mapper

from typing import Optional
from pydantic import BaseModel, EmailStr

metadata = MetaData()

user = Table('user', metadata,
            Column('id', Integer, primary_key=True),
            Column('name', String(50)),
            Column('email', String(100)),
            Column('fullname', String(50)),
            Column('nickname', String(12))
        )



class User(BaseModel):
    email: Optional[EmailStr] = None
    is_active: Optional[bool] = True
    is_superuser: Optional[bool] = False
    full_name: Optional[str] = None
    def __init__(self, name, email, fullname, nickname):
        self.name = name
        self.email = email
        self.fullname = fullname
        self.nickname = nickname


mapper(User, user)

```

**Comment by sandys** [comment]
Added a working code example with inheritance and sqlalchemy mixins .
We need to use pydantic support for python dataclasses rather than BaseModel. If we use plain BaseModel then sqlalchemy does not play nice with the classes.

```

from sqlalchemy import Table, Text, MetaData, Column, Integer, String, ForeignKey, create_engine
from sqlalchemy.orm import mapper, relationship, sessionmaker

from typing import Optional
from pydantic import  EmailStr
from pydantic.dataclasses import dataclass as py_dataclass
from dataclasses import dataclass

metadata = MetaData()

audit_mixin = Table('audit_mixin', metadata,
                     Column('id',Integer, primary_key=True),
                      Column('performed_by',Integer, nullable=True))

user = Table('user', metadata,
            Column('user_id', Integer, primary_key=True),
            Column('name', String(50)),
            Column('email', String(100)),
            Column('fullname', String(50)),
            Column('nickname', String(12))
        )


@dataclass
class AuditMixin:
    id: int
    performed_by: int 

@py_dataclass
class AuditMixinPy(AuditMixin):
    pass

@dataclass
class User(AuditMixin):
    user_id :int
    identity :str
    identity_type :str
    row_status :str
    comments :str

@py_dataclass
class UserPy(User):
    pass


engine = create_engine("sqlite:///", echo=False)
metadata.create_all(engine)

session = sessionmaker(engine)()

mapper(AuditMixin, audit_mixin)
mapper(User,user)

u = User(100,123,1, "a","dfd","dfdf","dfdd")
u.id
u.performed_by
session.add(u)
session.commit()

u = User("sss",123,1, "a","dfd","dfdf",1)
u = UserPy("sss",123,1, "a","dfd","dfdf",1)


```

**Comment by sandys** [comment]
fastapi can include a feature  (like Pyrolysis python server) that automatically maps dataclasses.

https://github.com/fmerlin/pyrolysis/blob/2edf1094466b09b4c5d1ab93f5cbc2bb01cc73c3/example/server.py#L15

**Comment by bubthegreat** [comment]
Was just trying to solve a similar problem for GINO and found pydantic's dynamic model generation handy.  I haven't played much with sqlalchemy directly, but it should be straightforward to implement recursive checking based on model relations, etc. similar to how tortoiseorm has done.

```

def get_pmodel(model):
    # Get all the keys to scan through.
    keys = [str(key) for key in model.__dict__.keys()]
    # Assumption that may not be valid, but don't look at ones with _ in them.
    valid_keys = [key for key in keys if not key.startswith('_')]

    # Allow exclusions of model attributes from the pydantic model.
    if hasattr(model, '__excluded__'):
        valid_keys = [key for key in valid_keys if key not in model.__excluded__]

    # This may be unique to GINO where the python type is on a column, but
    # It would be easy enough to make a reference table for this rather than
    # pulling it directly from the model column.
    field_definitions = {}
    for key in valid_keys:
        col = getattr(model, key)
        col_type = col.type.python_type
        # Don't forget ellipses after this, or pydantic won't create
        # the schema and validators properly.
        field_definitions[key] = (col.type.python_type,...)

    # Create our pydantic model
    pmodel = create_model(model.__name__, **field_definitions)

    return pmodel 
```

This can probably be simplified even further by just adding the _validator method to the base class and checking it properly, but the logic is sound.  Note that when not creating it dynamically with the ellipses it won't create any of the required schemas/validators.

```
class User(db.Model):
    __tablename__ = 'users'
    
    name = db.Column(db.String())
```

```
PUser = get_pmodel(User)
```

```
print(PUser.schema_json(indent=2))
{
  "title": "User",
  "type": "object",
  "properties": {
    "name": {
      "title": "Name",
      "type": "string"
    },
    "id": {
      "title": "Id",
      "type": "integer"
    }
  },
  "required": [
    "name",
    "id"
  ]
}
```

**Comment by bubthegreat** [comment]
Circling back to this - I set up a working "autocrud" setup that seems to be working fairly well using GINO as the ORM that supports alembic migrations: https://gitlab.com/bubthegreat/cookiecutter-gino-fastapi

It's dirty and needs to be cleaned up, but you can see how I've [integrated the pydantic model generation into a model base mixin](https://gitlab.com/bubthegreat/cookiecutter-gino-fastapi/-/blob/master/%7B%7Bcookiecutter.python_slug%7D%7D/src/%7B%7Bcookiecutter.python_slug%7D%7D/models/__init__.py#L31) that adds some methods for this, and then [used those mixin methods](https://gitlab.com/bubthegreat/cookiecutter-gino-fastapi/-/blob/master/%7B%7Bcookiecutter.python_slug%7D%7D/src/%7B%7Bcookiecutter.python_slug%7D%7D/utils/crud.py#L23) to generate an "autocrud" generation for all your models with just a few lines of code.

**Comment by YuriiMotov** [comment, marked as answer]
Looks like [SQLModel](https://sqlmodel.tiangolo.com/) might be helpful here
